google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/DefaultCredentialProvider.java

```diff
@@ -256,19 +258,45 @@ private GoogleCredential getCredentialUsingWellKnownFile(
     }
   }
 
-
-  private boolean runningOnAppEngine() {
-    if (getEnvEquals("GAE_ENV", "standard")) {
+  private boolean useGAEStandardAPI() {
+    if (getEnvEquals("GAE_ENV", "standard") && getEnvEquals("GAE_RUNTIME", "java7")) {
       return true;
     }
-    if (getEnvEquals("GAE_RUNTIME", "java7")) {
-      return true;
+
+    Class<?> systemPropertyClass;
+    try {
+      systemPropertyClass = forName("com.google.appengine.api.utils.SystemProperty");
+    } catch (ClassNotFoundException expected) {
+      // SystemProperty will always be present on App Engine.
+      return false;
     }
-    if (getEnvEquals("GAE_RUNTIME", "java8")) {
-      return true;
+    Exception cause = null;
+    Field environmentField;
+    try {
+      environmentField = systemPropertyClass.getField("environment");
+      Object environmentValue = environmentField.get(null);
+      Class<?> environmentType = environmentField.getType();
+      Method valueMethod = environmentType.getMethod("value");
+      Object environmentValueValue = valueMethod.invoke(environmentValue);
+      return (environmentValueValue != null);
+    } catch (NoSuchFieldException ignored) {
+      // If the field does not exist then we treat it as false.
+    } catch (SecurityException exception) {
```

Comment:
We're still stuck on Java 5, right? Otherwise you could use a multi-catch here:

```java
} catch (SecurityException | IllegalArgumentException | IllegalAccessException | NoSuchMethodException
    | InvocationTargetException exception) {
  cause = exception;
}
```

Reply:
Correct, unfortunately :/